src/rust/engine/fs/store/src/snapshot_ops.rs

```diff
@@ -0,0 +1,763 @@
+// Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
+// Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+use crate::{snapshot::osstring_as_utf8, Snapshot, Store};
+
+use async_trait::async_trait;
+use bazel_protos::remote_execution as remexec;
+use boxfuture::{BoxFuture, Boxable};
+use fs::{
+  ExpandablePathGlobs, GitignoreStyleExcludes, PathGlob, PreparedPathGlobs, DOUBLE_STAR_GLOB,
+  SINGLE_STAR_GLOB,
+};
+use futures::compat::Future01CompatExt;
+use futures::future::{self as future03, TryFutureExt};
+use futures01::future::{self, Future};
+use glob::Pattern;
+use hashing::{Digest, Fingerprint};
+use indexmap::{self, IndexMap, IndexSet};
+
+use std::convert::From;
+use std::iter::Iterator;
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+
+#[derive(Clone, Eq, PartialEq, Hash, Debug)]
+pub enum SnapshotOpsError {
+  String(String),
+  DigestMergeFailure(String),
+  GlobMatchError(String),
+}
+
+impl From<String> for SnapshotOpsError {
+  fn from(err: String) -> Self {
+    Self::String(err)
+  }
+}
+
+#[derive(Debug, Clone)]
+pub struct SubsetParams {
+  pub globs: PreparedPathGlobs,
+}
+
+#[derive(Clone)]
+pub enum MergeBehavior {
+  NoDuplicates,
+  LinearCompose(SubsetParams),
+}
+
+impl MergeBehavior {
+  pub fn get_filter_message(&self) -> String {
```

Comment:
AFAICT, this method is unused. In general, for methods/fields that are only intended to be used within a single file, it's good to remove `pub` will allow the dead code detection to kick in.

Reply:
Thank you for the tip, and yes this is indeed dead code!